### DISCUSSION TITLE
Export Templates - missing some fields in export

### DISCUSSION THREAD

**Discussion by Joelp** [netbox-community/netbox#11762] [Help Wanted!] [answered]
I created export template  for Exporting cables. This is my jinja:

```
side_a_device, side_a_type, side_a_name, side_b_device, side_b_type, side_b_name, status, tenant, color, length, length_unit, id
{%- for cable in queryset %}
{{ cable.side_a_device }}, {{ cable.side_a_type }}, {{ cable.side_a_name }}, {{ cable.side_b_device }}, {{ cable.side_b_type }}, {{ cable.side_b_name }}, {{ cable.status }}, {{ cable.tenant }}, {{ cable.color }}, {{ cable.length }}, {{ cable.length_unit }}, {{ cable.id }}
{%- endfor %}
```

It returns no values for some fields:

```
side_a_device, side_a_type, side_a_name, side_b_device, side_b_type, side_b_name, status, tenant, color, length, length_unit, id
, , , , , , connected, Rack S07, 111111, 5.00, m, 15
, , , , , , connected, Rack S07, 111111, 5.00, m, 16
, , , , , , connected, Rack S07, 111111, 5.00, m, 17
, , , , , , connected, Rack S07, 111111, 5.00, m, 18
```

I should return values for `side_a_device`, `side_a_type`, `side_a_name`, `side_b_device`, `side_b_type`, `side_b_name` too

Some fields are empty even if the cable is connected properly.

**Comment by candlerb** [comment, marked as answer]
There is no attribute `side_a_device` on the Cable model. And I'm afraid that the CSV import columns can't be used as a guide (nor the CSV export for that matter).

To explore the available attributes, use [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/).

```
source /opt/netbox/venv/bin/activate
python /opt/netbox/netbox/manage.py nbshell

>>> c = Cable.objects.first()
>>> c.side_a_device
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'Cable' object has no attribute 'side_a_device'
```

`help(Cable)` gives the detailed info, or `dir(c)` on an individual cable object, which you can then filter:

```
>>> [n for n in dir(c) if "a_" in n]
['a_terminations']
```

Note that recent versions of Netbox allow a single cable to connect to multiple interfaces on multiple devices at each end.  But if you don't use this capability, then `a_terminations[0]` and `b_terminations[0]` may be what you want.

```
>>> c.a_terminations
[<FrontPort: F16>]
>>> c.a_terminations[0].__class__
<class 'dcim.models.device_components.FrontPort'>
>>> c.a_terminations[0].name
'F16'
```

**Comment by Joelp** [comment]
I really appreciate the help! 

It was really hard to find everything what I need, but looks it works now. It will be nice if this will be well documented.  And the absolute best if import fields correspond with export fields.

However this is my export template which is usable for import.

```
id, side_a_device, side_a_type, side_a_name, side_b_device, side_b_type, side_b_name, type, status, tenant, color, length, length_unit
{%- for cable in queryset %}
{{ cable.id }}, {{ cable.a_terminations[0].device.name }}, {{cable.a_terminations[0].cable_terminations.content_type.app_label}}.{{cable.a_terminations[0].cable_terminations.content_type.model}}, {{ cable.a_terminations[0].name }}, {{ cable.b_terminations[0].device.name if cable.b_terminations[0].device is defined }}, {{cable.b_terminations[0].cable_terminations.content_type.app_label}}.{{cable.b_terminations[0].cable_terminations.content_type.model}}, {{ cable.b_terminations[0].name }}, {{ cable.type }},{{ cable.status }}, {{ cable.tenant }}, {{ cable.color }}, {{ cable.length }}, {{ cable.length_unit }}
{%- endfor %}

```

**Reply by candlerb** [reply]
You reminded me I'd done something similar a while back:

```
side_a_device,side_a_type,side_a_name,side_b_device,side_b_type,side_b_name,type,status,label,color,length,length_unit
{%- for c in queryset %}
{%- if c.termination_a_type.app_label != 'circuits' and c.termination_b_type.app_label != 'circuits' %}
{{ c.termination_a.device.name }},{{ c.termination_a_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_a.name }},{{ c.termination_b.device.name }},{{ c.termination_b_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_b.name }},{{ c.type }},{{ c.status }},"{{ c.label }}",{{c.color}},{{ c.length|replace("None","") }},{{ c.length_unit }}
{%- endif %}
{%- endfor %}
```

But this no longer works due to the cable model changes, so thank you for providing the fixes!

It's pretty ugly having to go via `.cable_terminations.content_type.(app_label|model)` just to get the model name :-(  I thought it might be possible to do this (to avoid touching the database again):

```
>>> c = Cable.objects.first()
>>> t = c.a_terminations[0]
>>> t.__class__._meta.app_label
'dcim'
>>> t.__class__._meta.model_name
'frontport'
```

Unfortunately this doesn't work in an export template, as it gives an error:

> ```
> There was an error rendering the selected export template (Cables for CSV re-import): access to attribute '__class__' of 'FrontPort' object is unsafe.
> ```

Grr.  So the best I can do is to simplify it slightly with some variable assignments:

```
side_a_device,side_a_type,side_a_name,side_b_device,side_b_type,side_b_name,type,status,label,color,length,length_unit
{%- for c in queryset %}
{%- set act = c.a_terminations[0].cable_terminations.content_type %}
{%- set bct = c.b_terminations[0].cable_terminations.content_type %}
{%- if act.app_label != 'circuits' and bct.app_label != 'circuits' %}
{{ c.a_terminations[0].device.name }},{{ act.app_label }}.{{ act.model }},{{ c.a_terminations[0].name }},{{ c.b_terminations[0].device.name }},{{ bct.app_label }}.{{ bct.model }},{{ c.b_terminations[0].name }},{{ c.type }},{{ c.status }},"{{ c.label }}",{{c.color}},{{ c.length|default("", True) }},{{ c.length_unit }}
{%- endif %}
{%- endfor %}
```

**Reply by Joelp** [reply]
And it's better for reading. Thanks for hint.

**Comment by Bottman99** [comment]
Hello all,

thanks for your thoughts/code on the cable exporter in the newer versions of netbox.

I am not so in with Pyhon/Django/Json/jaml....

How would the code have to be adapted if I have multi terminations for the cables? That doesn't quite make sense to me, and for you guys it's probably just a small consideration.
Simply to remove the "[0]"... it is not so easy, - does not work

**Reply by candlerb** [reply]
The first question is, "how would you want to represent multiple terminations in a CSV anyway?"

As far as I know, Netbox's CSV cable _import_ doesn't support multiple terminations, and hence any _export_ format you choose won't be usable to re-import.

The internal model has separate 'Cable' and 'CableTermination' objects - but 'CableTermination' is not exposed to the user.

**Reply by Bottman99** [reply]
Thanks for your quick reply, i.e. there is no way to get multi-termination cable into netbox via script? Hidden for the user did you mean?
Only the manual way? This is hardly feasible for us.

I could already put together a code that spit out a CSV containing multi-termination cables. But makes only half sense if the import of those is excluded. Only for documentation purposes for other departments / circuit diagram creation it could be useful.

**Reply by candlerb** [reply]
> there is no way to get multi-termination cable into netbox via script?

Via Python script (including "custom script"): yes, you have direct access to the underlying Django models.

Via REST API: not tested, but I think it should work. I see separate endpoints for `/dcim/cable-terminations/` and `/dcim/cables/` in the openapi docs, and there are separate arrays for `a_terminations` and `b_terminations` in dcim_cables_create.

Via CSV import: I don't think so.

**Reply by msmithaxesssystemscouk** [reply]
Hi, sorry for the thread bump, thanks to everyone above for the pointers.  

This is nearly but not quite working for me.  

I found that the template above put all records on one line instead of one record per line, also instead of dcim.interface and dcim.powerport I get dcim"."interface and dcim"."powerport, and the extra " screw up the import.

Here's my template -

```
"label","status","type","color","length","length_unit","side_a_device","side_a_type","side_a_name","side_b_device","side_b_type","side_b_name","tenant" {% for c in queryset %} {%- set act = c.a_terminations[0].cable_terminations.content_type %} {%- set bct = c.b_terminations[0].cable_terminations.content_type %} 
"{{ c.label }}","{{ c.status }}","{{ c.type }}","{{ c.color }}","{{ c.length|default("", True) }}","{{ c.length_unit }}","{{ c.a_terminations[0].device.name }}","{{ act.app_label }}"."{{ act.model }}","{{ c.a_terminations[0].name }}","{{ c.b_terminations[0].device.name }}","{{ bct.app_label }}"."{{ bct.model }}","{{ c.b_terminations[0].name }}","{{ c.tenant }}" 
{%- endfor %}

```

and some sample output -

```
"","connected","cat6a","2196f3","3.00","m","SAN01-4","dcim"."interface","mgmt0","N9300SW1","dcim"."interface","gi1/1/16","Tenant" 
"","connected","cat6a","2196f3","3.00","m","|SAN01","dcim"."interface","mgmt1","N9300SW2","dcim"."interface","gi1/1/16","Tenant"

```

**Reply by candlerb** [reply]
> the extra " screw up the import.

That's a bug in the template. Simply remove the superfluous quotes, i.e. replace

```
"{{ act.app_label }}"."{{ act.model }}"
```

with

```
"{{ act.app_label }}.{{ act.model }}"
```

**Reply by msmithaxesssystemscouk** [reply]
Ha ha, well spotted Brian!  Working like a charm now :-)
